### ISSUE TITLE
Match Included Mappings with OS Package Versions

### ISSUE THREAD

**Issue by DaAwesomeP** [include-what-you-use/include-what-you-use#1135]
I do not know if the Debian packages or other OS packages are maintained by the maintainers of this project or external maintainers; please close this if it is not relevant.

IWYU for Debian 11 Bullseye is v0.15 (https://packages.debian.org/bullseye/iwyu). The version of Boost included with Debian 11 is Boost v1.74, but the mapping files associated with IWYU v0.15 are only for Boost v1.64. Debian is just one example; I have not checked other distributions.

Right now the Boost mappings are tied to the IWYU version, but I am not sure how practical this is because of OS distribution inconsistencies like the one mentioned above. If these mappings were auto-generated at package build time, then they could be matched to the correct version provided by the OS (or manually provided by the user's project). I see in various other issues that various people are working on this but it seems there is not a concrete solution yet.

Another option would be to release a patch (either v0.15.1 or Debian build 8.15-3) that provides new mappings as they become available (or once for Debian 11 to fix this inconsistency). If the mapping syntax changed between versions of IWYU then this could become difficult to maintain, but I do think it is possible since the mappings are independent of the actual IWYU codebase.

It would also be possible to release separate packages for each mapping (i.e. `iwyu-mapping-boost`), but I think this is much more cumbersome and it doesn't solve the issue with the current OS versions that do not allow new packages in stable.

Another related issue is that of course a mapping for every version of Boost is not available, and knowing the path to the mapping file requires knowledge of which Boost version mappings are included with your version of IWYU. It is a bit confusing that `boost-all.imp` is actually out of date and not just a symlink to the most recent version.

At the very least, it would be good to look at the current library versions supported by all distributions with officially supported IWYU packages and ensure the required mappings are available.

**Comment by kimgr** [comment]
I think it's fair to think of the mappings as "contrib" content. We include them, but don't really maintain them (because it would require that we used and installed all the weird libraries people like :)).

Instead, we try to provide stable scripts so that you can regenerate these library mappings ad hoc from a representative environment for your project. These scripts are in the `mapgen` directory in the IWYU source tree -- I'm actually not sure if packagers install those. Maybe we should add them to an install step so they end up in the install root somewhere.

**Comment by kimgr** [comment]
(I was working on one for Boost at one point, but got distracted. Can see if I can revive it if you want to give it a go.)

**Comment by DaAwesomeP** [comment]
Yeah it would be great if the OS-provided packages could include these scripts! And it would be even better if we could get them to bundle a pre-generated run for the OS-provided library versions. I agree that is much better than bundling pre-generated mappings to arbitrary library versions from the IWYU source tree.

Ill take a look at the Debian side of package generation soon and also the Boost scripts!

**Comment by kimgr** [comment]
That would create a weird dependency between packages, i.e. the IWYU package would need access to Boost, Qt, etc. I think a better partitioning of responsibilities is to let the respective library packages ship with IWYU mappings, but that's a little presumptuous (IWYU is maybe not _that_ important :)).

So the right trade-off, to me, seems to be to make sure IWYU packages come with the mapgen scripts, so that users can solve their own problems in their own environment.

**Comment by DaAwesomeP** [comment]
I see. I suppose it would only be a development/build dependency, but that is not great. Including the mapgen scripts does seem like the best option then.

I would be interested in seeing if there is a way to automate such scripts so that they can be as painless as possible to use. I wonder if there would be a way to allow the CMake IWYU integration to compile its own cache of magen results using the libraries sources included in the CMake project. A user could simply then install IWYU, add a few lines to their `CMakeLists.txt` to enable the mapgen scripts they need, and then it would take care of the rest when CMake configures itself. It would be entirely platform independent and remain a portable configuration solution. Since the mappings are generated by the user, I think it makes sense to store them in the CMake cache.

This would require CMake to know the location of the magen scripts which may not be consistent on all OSes. This is an issue I have with the currently precompiled mappings. @kimgr what do you think about including a command like `iwyu-mapgen` which would only add one command to the `PATH` but allow for scripts to more easily take advantage of the mapgen scripts? Could be a simple Bash/shell wrapper with commands like:

```
$ iwyu-mapgen --help
$ iwyu-mapgen --version
$ iwyu-mapgen --list-available
$ iwyu-mapgen --list-available --machine-readable
$ iwyu-mapgen boost <args>
$ iwyu-mapgen cpython <args>
$ iwyu-mapgen qt <args>
```

Ideally there would be a standard minimum set of supported arguments for each mapgen script and then any optional flags specific to each generator. A simple `--machine-readable` or similar flag might allow tools to detect which mapgen scripts are available.

**Comment by kimgr** [comment]
To be honest, all I see is:
```
$ iwyu-mapgen --version
# not sure why this is necessary, but OK:
# iwyu-mapgen-boost.py --version, etc

$ iwyu-mapgen --list-available
# ls mapgen/

$ iwyu-mapgen --list-available --machine-readable
# ls mapgen/

$ iwyu-mapgen boost <args>
# $ mapgen/iwyu-mapgen-boost.py <args>

$ iwyu-mapgen cpython <args>
# mapgen/iwyu-mapgen-cpython.py <args>

$ iwyu-mapgen qt <args>
# mapgen/iwyu-mapgen-qt.py <args>
```

Integrating this into CMake only helps the CMake use case, not `iwyu_tool.py` or direct `include-what-you-use` invocations.

Let's focus on small steps and get the mapgen scripts included in `cmake install` so that packagers can/will include them.
